xdsl/frontend/passes/desymref.py

```diff
@@ -0,0 +1,389 @@
+#
+# 1. variable declaration, e.g. `a: i32 = 0` maps to `declare @a`
+# 2. variable use, e.g. `... = a` maps to `... = fetch @a`
+# 3. variable assignement, e.g. `a = ...` maps to `update @a ...`
+#
+# With these, it is relatively straightforward to lower any Python program to
+# xDSL (taking care of scoping rules, of course). For example, the code above
+# would become (in a slightly abused xDSL syntax):
+#
+# ```
+# func foo() -> i32 {
+#   declare @a
+#   update @a with 0
+#   for 0 to 100 {
+#     t1 = fetch @a
+#     t2 = add t1 2
+#     update @a t2
+#   }
+#   t3 = fetch @a
+#   return t3
+# }
+# ```
+#
+# Note that while insertion of symref operations is easy, the generated xDSL is
+# not fully in SSA form. For example, blocks in region should pass values
+# between each other via block arguments, instead of symref calls. Similarly,
+# some operations like scf.if or affine.for should yield a value.
+#
+#
+# Desymrefication pass
+# ====================
+#
+# In this file, we implement desymrefication pass - it goes over generated xDSL
+# and removes all symref operations. To describe how the pass works, first
+# recall that in xDSL an operation can contain regions which in turn can contain
+# CFGs of basic blocks. An operation passes control flow to a region and on exit
+# from the region the control flow returns to the operation. Then the operation
+# can transfer control flow to another region, etc.
+#
+# Desymrefication can be applied to regions and operations. First, consider a
+# simple case when all operations in the region do not have any nested regions.
+# This means that the region is simply a CFG. Otherwise, there are some
+# operations that contain regions. However, we can apply desymrefication to
+# these operations them first, reducing the problem to the the first case. In
+# pseudocode, that would look like:
+#
+# ```
+# desymref_region(region):
+#   for block in region.blocks:
+#     for op in block.ops:
+#       desymref_op(op)
+#     # at this point no op contains a symref operation
+#     # in the nested region.
+# ```
+#
+# Next, we describe how desymrefication actually works, on a high level.
+# Every region either declares a symbol or uses it from the parent region.
+# Let's consider each case separately.
+#
+# Case 1: Symbol is declared in this region. This is an easy case. We already
+# know that symbol is only used in this region and no other nested region uses
+# it. Therefore running any SSA-construction algorithm on the CFG is enough.
+#
+# Case 2: Symbol is not declared in this region. In this case, we only support
+# non-CFGs at the moment (i.e. single block regions). We observe that in general
+# every symbol can be simplified within a block to have at most one fetch (start
+# of the block) at most one update (end of the block). This means that using
+# specification of an operation we can promote these symbols outside.
+# TODO: Add op promotion.
+
+
+def is_definition(op: Operation) -> bool:
```

Comment:
Why do you have all these functions that just return an isinstance call? IMO, the only one contianing more info is `is_ues`, we should consider removing the others.

Reply:
Motivation was that if we change the `symref` dialect (rename things, introduce new ops, etc.) we would need less changes in the future to the algorithm. I am fine with using `isinstance` though :)